File: week1/LongestSubstringWithoutRepeatingCharacters/longestSubstringWithoutRepeatingCharacters.py
class Solution:
    @classmethod
    def longestSubstringWithoutRepeatingCharacters(self, s:str)->int:
        charSet = set()
        subWaler = 0
        maxLen = 0
        
        for walker in range(0, len(s)):
            if s[walker] not in charSet:
                charSet.add(s[walker])
                maxLen = len(charSet)
            else:
                while s[walker] in charSet:
                    charSet.remove(s[subWaler])
                    subWaler+=1
                charSet.add(s[walker])
        
        return maxLen

input1 = "abcabcbb"
output1 = Solution.longestSubstringWithoutRepeatingCharacters(input1)
print(output1)

File: week1/LongestSubstringWithoutRepeatingCharacters/longestSubstringWithoutRepeatingCharacters.cpp
#include <iostream>
#include <string>
#include <unordered_set>
using namespace std;

int longestSubstringWithoutRepeatingCharacters(string s);

int main(){
    string input = "abcabcbb";
    int output = longestSubstringWithoutRepeatingCharacters(input);
    std:: cout << output;
}

int longestSubstringWithoutRepeatingCharacters(string s){
    int maxLen = 0, subWalker=0;
    unordered_set<char> charSet;

    for(int i =0; i < s.length(); i++){
        if(!charSet.count(s[i])){
            charSet.insert(s[i]);
            maxLen = charSet.size();
        } else {
            while(charSet.count(s[i])){
                charSet.erase(s[subWalker]);
                subWalker++;
            }
            charSet.insert(s[i]);
        }
    }

    return maxLen;
}